src/persist-client/src/iter.rs

```diff
@@ -678,16 +688,36 @@ where
     /// Wait until data is available, then return an iterator over the next
     /// consolidated chunk of output. If this method returns `None`, that all the data has been
     /// exhausted and the full consolidated dataset has been returned.
-    #[allow(dead_code)]
     pub(crate) async fn next_chunk(
         &mut self,
         max_len: usize,
+        max_bytes: usize,
     ) -> anyhow::Result<Option<BlobTraceUpdates>> {
         self.trim();
         self.unblock_progress().await?;
-        let updates = self.iter().map(|mut i| i.next_chunks(max_len));
-        let updates = updates.map(|chunk| self.sort.updates_to_blob(chunk));
-        Ok(updates)
+
+        let Some(mut iter) = self.iter() else {
+            return Ok(None);
+        };
+
+        let parts = iter.parts.clone();
+
+        // Keep a running estimate of the size left in the budget, returning None after it's
+        // exhausted. Note that we can't use take_while here... it's important that we pass
+        // on every value we receive from the iterator.
```

Comment:
I might misunderstsand what this comment means, but I'm pretty sure we stop returning elements once the `budget == 0`? Here's an example from the [Rust Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=59bd7b0c534193baefe56ddd6a14a5d7)

Reply:
Yes, that's correct! The idea of the second half of this comment is that it would be bad for us to pull an element from the consolidating iter and not foward it along, so we can't use `take_while`. (Which drops the first non-matching element.) I'll rephrase.